base/reduce.jl

```diff
@@ -1246,3 +1184,84 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
     end
     return n
 end
+
+# simd optimization for min/max related reduction
+# general fallback
+_fast(op, x, y) = op(x, y)
+_fast(op) = (x, y) -> _fast(op, x, y)
+
+# The following are used in optimized `mapreduce_impl` for IEEEFloat.
+# They compare `0.0/-0.0` correctly.
+_fast(::typeof(min), x::T, y::T) where {T<:IEEEFloat} = ifelse(signbit(x-y), x, y)
+_fast(::typeof(max), x::T, y::T) where {T<:IEEEFloat} = ifelse(signbit(y-x), x, y)
+
+# Some help function for better vectorization
+# NaN check
+function _anynan(x::NTuple{N}) where {N}
+    r = false
+    for i in 1:N>>1
+        r |= isnan(x[i]) | isnan(x[i+N>>1])
+    end
+    r
+end
+# split then reduce for inputs whose length > 4
+_half_reduce(op, x::NTuple{N}) where {N} = _half_reduce(op, ntuple(i -> op(x[i], x[i+N>>1]), Val(N>>1)))
+_half_reduce(op, x::NTuple{4}) = reduce(op, x)
+
+# A help struct to make closure stable (We can't assume `T(x) isa T` in general)
+struct _StableApply{T1,T2} end
+(::_StableApply{T1,T2})(x) where {T1,T2} = T1(x)::T2
+_stable_apply(::Type{T1},::Type{T2}) where {T1,T2} = _StableApply{T1,T2}()
+_stable_apply(f, ::Any) = f # Function should be stable anyway
+
+# The optimized fallback for `IEEEFloat`'s min/max reduction. (LLVM handles `Integer` well)
+function mapreduce_impl(f::F, op::Union{typeof(max),typeof(min)},
+                        A::AbstractArrayOrBroadcasted, first::Int, last::Int) where {F}
+    T = @default_eltype A
+    Eltype = _return_type(f, Tuple{T})
+    # Call general fallback for non-IEEEFloat cases.
+    (isconcretetype(Eltype) && Eltype <: IEEEFloat) ||
+        return mapreduce_impl(f, op, A, first, last, pairwise_blocksize(f, op))
+    f′ = _stable_apply(f, Eltype)    # Keep stable if `f isa Type`
```

Comment:
I know I suggested `_stable_apply` but now that there's `isconcretetype` check, this is no longer required (it'd be a bug in the compiler if `f` returns other types).

Reply:
It is used here to avoid type-instability caused by #23618.
And force accurate inference for `minimum(Float64, a::Vector{Real})`.
Maybe we can wait for #40985? (I saw a 1.8 Milestone there)

Reply:
Ah, please forget my comment above. I thought you were calling `convert`.

In that case, I suppose you don't need to pass `Eltype`? It looks like `struct _StableApply{T1} end` should just work. I don't think we'd want to wait for #40985. I don't think the milestone there would be very effective since it is a simple enhancement (e.g., not a release-blocking bug). Also, we have workarounds like this in other places.

Reply:
`Eltype` was passed for type annotation. 
When I test `minimum(Float64, Vector{Real}(randn(4096)))`, `_returntype(Float64, Tuple{Real})` returns `Float64` but the infered type of `elf()` is `Any`.

In the retuned code, all `elf()` has been inlined, so I guess there's no need to worry about #23618. (Stability test has been added)

Reply:
Did you try `struct _StableApply{T1} end`? Workaround for #23618 does not require manual assertion.

Reply:
Hmm... I'm surprised this works. Did you check that `f(_)` is not a dynamic dispatch? I also would expect that `Eltype` is captured as a `DataType`...

https://github.com/JuliaLang/julia/blob/ac1fc8eeb8124d8e6734cd9e0c46d7cb5b23139e/base/reduce.jl#L1285

Reply:
I tried the following:
```julia
julia> using Test
julia> @inferred Base.mapreduce_impl(Float64,min,Real[1,2,3],1,3)
1.0
julia> @inferred Base.mapreduce_impl(Float64,min,[1,2,3],1,3)
1.0
```
Seems work well. I found the following in `@code_typed`
```julia
%108 = %new(Base.var"#292#295"{Int64, Base.var"#elf#293"{DataType, Vector{Real}, DataType}}, %104, %4)::Base.var"#292#295"{Int64, Base.var"#elf#293"{DataType, Vector{Real}, DataType}}
```
but in `@descend`:
```julia
     │     %402 = Base.arrayref(false, A, %401)::Realgetindex
     │     %403 = (Float64)(%402)::Any
     │            Core.typeassert(%403, Float64)   
```
I guess it's OK?

Reply:
`elf` has been totally droped in the lastest version